crates/wasmtime/src/component/func/typed.rs

```diff
@@ -1548,13 +1548,51 @@ unsafe impl<T: ComponentValue> ComponentValue for WasmList<T> {
     }
 }
 
+/// Round `a` up to the next multiple of `align`, assuming that `align` is a power of 2.
 #[inline]
 const fn align_to(a: usize, align: u32) -> usize {
     debug_assert!(align.is_power_of_two());
     let align = align as usize;
     (a + (align - 1)) & !(align - 1)
 }
 
+/// For a field of type T starting after `offset` bytes, updates the offset to reflect the correct
+/// alignment and size of T. Returns the correctly aligned offset for the start of the field.
+#[inline]
+pub fn next_field<T: ComponentValue>(offset: &mut usize) -> usize {
+    *offset = align_to(*offset, T::align());
+    let result = *offset;
+    *offset += T::size();
+    result
+}
+
+/// Verify that the given wasm type is a tuple with the expected fields in the right order.
+#[inline]
+pub fn typecheck_tuple(
```

Comment:
I think this probably won't be reused externally because it specifically matches the `Tuple` type, right?

Reply:
I think the wasm tuple type is the right thing for tuple-like structs and enum variants, so I'm expecting to use it for those. Does that seem right?

Reply:
In that you're mapping `struct Foo(u32, u32)` to `tuple<u32, u32>` in the component model? For that I'd probably just return an error on a struct looking like that since that's already exposed as tuples and there's not really much need to expose a different way to name tuples in Rust.

For enum variants are you supporting `enum Foo { A(u32, u32) }`? If so personally I'd prefer to reject that for now and stay strict in the derive to stay close to the component model representation if we can.

Depending how the whole ecosystem shakes out we can revisit in the future and add "sugar" if this sort of type structure ends up being common in wit

Reply:
Huh, I see. Sure, I can do it that way.

I still want to factor this function out of the `impl_component_ty_for_tuples` macro, even if it isn't public. It's very similar to the typechecking function for unions, and with the addition of names, both of those resemble the typechecking functions for records, enums, and variants. I think it'll be helpful to keep them all next to each other.

I've verified that rustc generates the same code on x86-64 release builds after refactoring the impl for tuples this way. (It inlines the helper function and unrolls the iterator, so the recursive typechecking calls are all direct.) So it seems to me that the main consideration here is whether the refactoring is easier to understand, and for me, getting code out of macros helps a lot in my ability to understand it.

Reply:
Ok that seems reasonable to me yeah. In general we don't have to worry too much about type-checking performance since that's typically front-loaded where possible off the critical path so it's ok to drop `#[inline]` on functions like this and otherwise just optimize it later if we really need to.

Reply:
Okay, I fixed `typecheck_tuple` to not be `pub`. Thanks for your review! Anything else blocking this from merge?